Unhelpful compiler advice when using context

I'm not sure if this is anyhow's fault, or the compiler's fault.

When using .context, I get the following error:

```
error[E0599]: no method named `context` found for type `std::result::Result<(), anyhow::error::Error>` in the current scope
  --> src/main.rs:83:13
   |
83 |             context(format!("failure in {} with nesting", c.name))?;
   |             ^^^^^^^ method not found in `std::result::Result<(), anyhow::error::Error>`
   |
   = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
   |
19 | use anyhow::context::Context;
```

That last line is awful advice, as that just leads to:

```
error[E0603]: module `context` is private
  --> src/main.rs:24:13
   |
24 | use anyhow::context::Context;
   |             ^^^^^^^
```

The real fix is: `use anyhow::Context`. I don't know how to fix this message, but I feel it would make anyhow more usable if it was fixed.

--- a/src/context.rs
+++ b/src/context.rs
@@ -1,69 +1,10 @@
-use crate::Error;
+use crate::{Context, Error};
 use std::convert::Infallible;
 use std::error::Error as StdError;
 use std::fmt::{self, Debug, Display};
 
 #[cfg(backtrace)]
 use std::backtrace::Backtrace;
-
-/// Provides the `context` method for `Result`.
-///
-/// This trait is sealed and cannot be implemented for types outside of
-/// `anyhow`.
-///
-/// # Example
-///
-/// ```
-/// use anyhow::{Context, Result};
-/// use std::fs;
-/// use std::path::PathBuf;
-///
-/// pub struct ImportantThing {
-///     path: PathBuf,
-/// }
-///
-/// impl ImportantThing {
-///     # const IGNORE: &'static str = stringify! {
-///     pub fn detach(&mut self) -> Result<()> {...}
-///     # };
-///     # fn detach(&mut self) -> Result<()> {
-///     #     unimplemented!()
-///     # }
-/// }
-///
-/// pub fn do_it(mut it: ImportantThing) -> Result<Vec<u8>> {
-///     it.detach().context("failed to detach the important thing")?;
-///
-///     let path = &it.path;
-///     let content = fs::read(path)
-///         .with_context(|| format!("failed to read instrs from {}", path.display()))?;
-///
-///     Ok(content)
-/// }
-/// ```
-///
-/// When printed, the outermost context would be printed first and the lower
-/// level underlying causes would be enumerated below.
-///
-/// ```console
-/// Error: failed to read instrs from ./path/to/instrs.jsox
-///
-/// caused by:
-///     No such file or directory (os error 2)
-/// ```
-pub trait Context<T, E>: private::Sealed {
-    /// Wrap the error value with additional context.
-    fn context<C>(self, context: C) -> Result<T, Error>
-    where
-        C: Display + Send + Sync + 'static;
-
-    /// Wrap the error value with additional context that is evaluated lazily
-    /// only once an error does occur.
-    fn with_context<C, F>(self, f: F) -> Result<T, Error>
-    where
-        C: Display + Send + Sync + 'static,
-        F: FnOnce() -> C;
-}
 
 mod ext {
     use super::*;
@@ -209,7 +150,7 @@ where
     }
 }
 
-mod private {
+pub(crate) mod private {
     use super::*;
 
     pub trait Sealed {}

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -183,7 +183,8 @@ mod kind;
 #[cfg(not(feature = "std"))]
 compile_error!("no_std support is not implemented yet");
 
-pub use crate::context::Context;
+use std::fmt::Display;
+
 pub use crate::error::{Chain, Error};
 
 /// `Result<T, Error>`
@@ -424,6 +425,65 @@ macro_rules! anyhow {
     ($fmt:expr, $($arg:tt)*) => {
         $crate::private::new_adhoc(format!($fmt, $($arg)*))
     };
+}
+
+/// Provides the `context` method for `Result`.
+///
+/// This trait is sealed and cannot be implemented for types outside of
+/// `anyhow`.
+///
+/// # Example
+///
+/// ```
+/// use anyhow::{Context, Result};
+/// use std::fs;
+/// use std::path::PathBuf;
+///
+/// pub struct ImportantThing {
+///     path: PathBuf,
+/// }
+///
+/// impl ImportantThing {
+///     # const IGNORE: &'static str = stringify! {
+///     pub fn detach(&mut self) -> Result<()> {...}
+///     # };
+///     # fn detach(&mut self) -> Result<()> {
+///     #     unimplemented!()
+///     # }
+/// }
+///
+/// pub fn do_it(mut it: ImportantThing) -> Result<Vec<u8>> {
+///     it.detach().context("failed to detach the important thing")?;
+///
+///     let path = &it.path;
+///     let content = fs::read(path)
+///         .with_context(|| format!("failed to read instrs from {}", path.display()))?;
+///
+///     Ok(content)
+/// }
+/// ```
+///
+/// When printed, the outermost context would be printed first and the lower
+/// level underlying causes would be enumerated below.
+///
+/// ```console
+/// Error: failed to read instrs from ./path/to/instrs.jsox
+///
+/// caused by:
+///     No such file or directory (os error 2)
+/// ```
+pub trait Context<T, E>: context::private::Sealed {
+    /// Wrap the error value with additional context.
+    fn context<C>(self, context: C) -> Result<T, Error>
+    where
+        C: Display + Send + Sync + 'static;
+
+    /// Wrap the error value with additional context that is evaluated lazily
+    /// only once an error does occur.
+    fn with_context<C, F>(self, f: F) -> Result<T, Error>
+    where
+        C: Display + Send + Sync + 'static,
+        F: FnOnce() -> C;
 }
 
 // Not public API.

